README.md
# Git stats

Provide MR/PR statistics for your Gitlab/GitHub projects

## Install

Run `npm install -g gitflows-stats`

### Prerequisite

**Generate a token to use the API:**

- [Gitlab](https://docs.gitlab.com/ee/user/profile/personal_access_tokens.html)
  - Have a token with enough privilege (`read_api`)
- [GitHub](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token#creating-a-personal-access-token-classic)
  - Generate a personal access token

## Usage

- `gitflows-stats --help` will print:

  ```shell
  Usage: index [options] [command]

  Options:
  -h, --help                                        display help for command

  Commands:
  gitlab [options] <token> <projectId> <period>     Provide merge requests statistics on a Gitlab project for a given period
  github [options] <token> <owner> <repo> <period>  Provide pull requests statistics on a GitHub project for a given period
  help [command]                                    display help for command
  ```

- the `<period>` parameter has an optional end and may be in the following format:

  - `2023-01-01,2023-01-31` will retrieve all the merged events between these 2 dates
  - `2023-01-01` will retrieve all the merged events between the given date and now

- `gitflows-stats gitlab <TOKEN> <PROJECT_ID> <PERIOD>` (<PERIOD> in the following format `2023-01-01,2023-01-31`) will print:
  ```json
  {
    "average": {
      "months": 0,
      "days": 4,
      "hours": 22,
      "minutes": 48,
      "seconds": 0
    },
    "total": {
      "merged": 15,
      "closed": 2,
      "opened": 3,
      "all": 20
    },
    "data": [
      [
        2023,
        [
          {
            "Week 9": 1
          },
          {
            "Week 10": 5
          },
          {
            "Week 11": 3
          },
          {
            "Week 12": 2
          },
          {
            "Week 13": 2
          },
          {
            "Week 14": 2
          }
        ]
      ]
    ]
  }
  ```
- `gitflows-stats gitlab <TOKEN> <PROJECT_ID> <PERIOD> --format html` (<PERIOD> in the following format `2023-01-01,2023-01-31`) will generate a `report` folder with an `index.html` where the command has been executed.

  **The report will automatically open after generation.**

  **example:**
  ![](documentation/chart_screenshot.png)

### Available output format

- **Console (Default):** print a lightweight statistics in JSON into the console
- **HTML:** generates a `HTML` file, opening automatically in your browser
- **CSV:** generates a `CSV` file with all raws data


CHANGELOG.md
# Changelog

## [0.3.1](https://github.com/bbougon/git-stats/compare/v0.3.0...v0.3.1) (2023-07-23)


### Bug Fixes

* Display dates instead of week numbers ([b978422](https://github.com/bbougon/git-stats/commit/b9784227c066e64840ce68f06323acd5b150458e))

## [0.3.0](https://github.com/bbougon/git-stats/compare/v0.2.8...v0.3.0) (2023-06-21)


### Features

* add estimated time and duration in front of each bar ([6f2cc34](https://github.com/bbougon/git-stats/commit/6f2cc341f51dfc135b1a19cab4954049d52e30f5))
* display 2 different progress bars for issues and events pagination ([a547c3c](https://github.com/bbougon/git-stats/commit/a547c3cfdbd91695cdaa05367816e0b8a56cba6b))

## [0.2.8](https://github.com/bbougon/git-stats/compare/v0.2.7...v0.2.8) (2023-06-18)


### Bug Fixes

* when a period end is given, the end date is set at the end of the day at 23:59:59 ([3f66b64](https://github.com/bbougon/git-stats/commit/3f66b64b1f63998fca41c6902d3fe07684d7d250))
* when statistics are generated for a given period (i.e with a given end date), issues that are closed after the end date are accounted as opened during this period ([521e5ad](https://github.com/bbougon/git-stats/commit/521e5ad3d93b5091e221c5a93082fc583319880e))
* when statistics are generated for a given period (i.e with a given end date), merge events that are closed after the end date are accounted as opened during this period ([f3e8430](https://github.com/bbougon/git-stats/commit/f3e84303b1cebe7bfaf569b6ef968b33aab68d49))
* when statistics are generated for a given period (i.e with a given end date), merged events that are merged after the end date are accounted as opened during this period ([2babb62](https://github.com/bbougon/git-stats/commit/2babb62cf1abb52272e4eb711e0384849dea00c5))

## [0.2.7](https://github.com/bbougon/git-stats/compare/v0.2.6...v0.2.7) (2023-06-16)


### Bug Fixes

* Add missing cumulative issues statistics that made the display in console failing ([f7c50e6](https://github.com/bbougon/git-stats/commit/f7c50e6ebb172e58255d23bcac4666e1ee90eb57))
* Add missing issues statistics that made the display in console failing ([a800708](https://github.com/bbougon/git-stats/commit/a800708e9252d5b490b1cf1a62b803505ac635b5))
* Add missing period issues statistics that made the display in console failing ([cb4e319](https://github.com/bbougon/git-stats/commit/cb4e3194b2a380383e61f934a62e844f64d4a027))
* display progress bar for github when no pagination available ([406f6e5](https://github.com/bbougon/git-stats/commit/406f6e5710cd1d06be91ac3d2ab7dcdf17cd6bfe))

## [0.2.6](https://github.com/bbougon/git-stats/compare/v0.2.5...v0.2.6) (2023-05-18)


### Features

* display issues statistics in HTML report ([a979476](https://github.com/bbougon/git-stats/commit/a979476e9f8a3c95186436e16bd92465c73e828a))

## [0.2.5](https://github.com/bbougon/git-stats/compare/v0.2.4...v0.2.5) (2023-05-13)


### Features

* display aggregated issues doughnut graph ([aabec3b](https://github.com/bbougon/git-stats/commit/aabec3b47fd774efe4da053239f8ee6651ea7579))
* generate cumulative statistics for issues ([1891402](https://github.com/bbougon/git-stats/commit/18914029091ad46dd5ab78e8d27159a9f8acba31))

## [0.2.4](https://github.com/bbougon/git-stats/compare/v0.2.3...v0.2.4) (2023-05-07)


### Features

* aggregate statistics for issues ([d2022fa](https://github.com/bbougon/git-stats/commit/d2022fa7377e441cf7a89895d44a1b1e4f6e68b8))
* Display aggregated issues events oin issue panel ([bfd4f04](https://github.com/bbougon/git-stats/commit/bfd4f04ea9934a8c813ea6d8e9dffab85ca3a137))

## [0.2.3](https://github.com/bbougon/git-stats/compare/v0.2.2...v0.2.3) (2023-05-01)


### Features

* display cumulative statistics in console. ([5e32935](https://github.com/bbougon/git-stats/commit/5e329354c501f43b83694cbe1800657002cb2d6e))

## [0.2.2](https://github.com/bbougon/git-stats/compare/v0.2.1...v0.2.2) (2023-04-22)


### Features

* display burnup in HTML format ([bf43f0c](https://github.com/bbougon/git-stats/commit/bf43f0c6e162cfb4a2e61fa30f6cf57e135fc47c))
* generate cumulative statistics ([c4618e6](https://github.com/bbougon/git-stats/commit/c4618e6966c199f37d87506ac80b1323744b596f))


### Bug Fixes

* trend calculation ([872c0ad](https://github.com/bbougon/git-stats/commit/872c0adc45e07079a09417aa4ee9b038178444fc))

## [0.2.1](https://github.com/bbougon/git-stats/compare/v0.2.0...v0.2.1) (2023-04-19)


### Features

* Add median time and average time per period: ([3bfae34](https://github.com/bbougon/git-stats/commit/3bfae34c9cc538dd8c684cd6a77daa8a7426c4fc))
* Display months and weeks events ([3120c72](https://github.com/bbougon/git-stats/commit/3120c72b838e57b8952066471ec088e2eb8bdf0a))

## [0.2.0](https://github.com/bbougon/git-stats/compare/v0.1.4...v0.2.0) (2023-04-12)


### Features

* end period date is now optional ([224c4a8](https://github.com/bbougon/git-stats/commit/224c4a8e2e294ff7b422e7afe54c8ab441b85dc9))


### Bug Fixes

* chart display weeks instead of months ([2134be3](https://github.com/bbougon/git-stats/commit/2134be38849e8adeffc1803b45066d20293b140d))

## [0.1.4](https://github.com/bbougon/git-stats/compare/v0.1.3...v0.1.4) (2023-04-12)


### Bug Fixes

* release process ([248581d](https://github.com/bbougon/git-stats/commit/248581de65916161efdce538de6b5ba7ec471d45))


index.ts
#!/usr/bin/env node
import { Command, program } from "commander";
import { add, parseISO } from "date-fns";
import { ConsoleWriter } from "./src/infrastructure/writer/ConsoleWriter.js";
import { CSVWriter } from "./src/infrastructure/writer/CSVWriter.js";
import { gitStatistics, StatisticsAggregate } from "./src/statistics/GitStatistics.js";
import { MergeEvent } from "./src/statistics/merge-events/MergeEvent.js";
import { GitlabEventParameters } from "./src/statistics/Gitlab.js";
import { PullRequestsStatsParameter } from "./src/statistics/Github.js";
import { ProgressBar } from "./src/infrastructure/progress-bar/ProgressBar.js";
import * as chalk from "./src/infrastructure/progress-bar/Chalk.js";
import { HTTPError } from "./src/infrastructure/repository/EventHTTPRepository.js";
import { Repositories } from "./src/statistics/Repositories.js";
import { IssueEvent } from "./src/statistics/issues/Issues.js";
import { MergedRequestHTTPGitlabRepository } from "./src/infrastructure/repository/gitlab/MergeRequestHTTPGitlabRepository.js";
import { PullRequestHTTPGithubRepository } from "./src/infrastructure/repository/github/PullRequestHTTPGithubRepository.js";
import { IssueHTTPGitlabRepository } from "./src/infrastructure/repository/gitlab/IssueHTTPGitlabRepository.js";
import { EventRepository } from "./src/statistics/EventRepository.js";
import { IssueHTTPGithubRepository } from "./src/infrastructure/repository/github/IssueHTTPGithubRepository.js";
import { HTMLWriter } from "./src/infrastructure/writer/html/HTMLWriter.js";

type Period = {
  start: Date;
  end: Date;
};

const commaSeparatedList = (list: string): Period => {
  const period = list.split(",");
  if (period.length === 1) {
    return {
      end: new Date(),
      start: parseISO(period[0]),
    };
  }
  return { end: add(parseISO(period[1]), { hours: 23, minutes: 59, seconds: 59 }), start: parseISO(period[0]) };
};

const writer = (format: string): Writer => {
  if (format === "html") {
    return new HTMLWriter("./");
  }
  if (format === "csv") {
    return new CSVWriter("./");
  }
  return new ConsoleWriter();
};

export interface Writer {
  write(stats: StatisticsAggregate): void;
}
export type RequestParameters = {
  fromDate: Date;
  toDate: Date;
};
type CommandParameters = {
  requestParameters: RequestParameters;
  options: any;
  token: string;
};

function isHTTPError(reason: HTTPError | string | never): reason is HTTPError {
  return (reason as HTTPError).rationale !== undefined;
}

const gitlabCommand = program
  .command("gitlab")
  .description("Provide merge requests statistics on a Gitlab project for a given period")
  .argument("<token>", "your gitlab API token")
  .argument("<projectId>", "gitlab project id for which you want to have statistics")
  .argument(
    "<period>",
    "the period you want to analyse (ISO formatted date with\nthe start period, e.g: 2021-11-02\nor separated by comma, e.g: 2021-11-02,2021-11-03 to retrieve merged events between these 2 dates)",
    commaSeparatedList
  );

const githubCommand = program
  .command("github")
  .description("Provide pull requests statistics on a GitHub project for a given period")
  .argument("<token>", "your github API token")
  .argument("<owner>", "The account owner of the repository. The name is not case sensitive")
  .argument("<repo>", "The name of the repository. The name is not case sensitive.")
  .argument(
    "<period>",
    "the period you want to analyse (ISO formatted date with\nthe start period, e.g: 2021-11-02\nor separated by comma, e.g: 2021-11-02,2021-11-03 to retrieve merged events between these 2 dates)",
    commaSeparatedList
  );

const proceedCommand = (
  command: Command,
  commandParameters: (...args: any[]) => CommandParameters,
  repositories: (token: string) => Repositories
) => {
  command
    .option(
      "-f, --format <writer>",
      "format to display the stats (default json in console), available formats are html, csv and console (default)",
      writer,
      new ConsoleWriter()
    )
    .action((...args: any[]) => {
      const parameters = commandParameters(...args);
      Repositories.initialize(repositories(parameters.token));
      gitStatistics(parameters.requestParameters)
        .then((stats) => {
          parameters.options.format.write(stats);
          ProgressBar.progressBar().clear();
        })
        .catch((reason: HTTPError | string | never) => {
          ProgressBar.progressBar().clear();
          const error = new chalk.chalk.Chalk().red;
          let errorMessage = reason;
          if (isHTTPError(reason)) {
            const additionalInfo = Object.entries(reason.additionalInfo as object).reduce(
              (accumulator, currentValue) => accumulator.concat(`- ${currentValue[0]}: ${currentValue[1]}\n\t`),
              ""
            );
            errorMessage = `${reason.rationale}.\nBelow are some additional information:\n\t${additionalInfo}`;
          }
          console.log(error(`Something wrong happened:\n${errorMessage}`));
        });
    });
};

proceedCommand(
  gitlabCommand,
  (token: string, projectId: number, period: Period, options) => ({
    requestParameters: {
      fromDate: period.start,
      projectId: projectId,
      toDate: period.end,
    } as GitlabEventParameters,
    options,
    token,
  }),
  (token: string) =>
    new (class extends Repositories {
      protected getIssueEventRepository(): EventRepository<IssueEvent> {
        return new IssueHTTPGitlabRepository(token);
      }

      protected getMergeEventRepository(): EventRepository<MergeEvent> {
        return new MergedRequestHTTPGitlabRepository(token);
      }
    })()
);

proceedCommand(
  githubCommand,
  (token: string, owner: string, repo: string, period: Period, options) => ({
    requestParameters: {
      repo,
      owner,
      fromDate: period.start,
      toDate: period.end,
    } as PullRequestsStatsParameter,
    options,
    token,
  }),
  (token: string) =>
    new (class extends Repositories {
      protected getIssueEventRepository(): EventRepository<IssueEvent> {
        return new IssueHTTPGithubRepository(token);
      }

      protected getMergeEventRepository(): EventRepository<MergeEvent> {
        return new PullRequestHTTPGithubRepository(token);
      }
    })()
);

program.parse();
